diff --git a/index.js b/index.js
index ed4b6d5b6cbabee76c879c6eedcc7bf6ba46d38d..8752893ca4986cf05103a46c438223b7ee4e1907 100644
--- a/index.js
+++ b/index.js
@@ -441,7 +441,7 @@ add(paths_, _origAdd, _internal) {
 
   if (this.options.useFsEvents && this._fsEventsHandler) {
     if (!this._readyCount) this._readyCount = paths.length;
-    if (this.options.persistent) this._readyCount *= 2;
+    if (this.options.persistent) this._readyCount += paths.length;
     paths.forEach((path) => this._fsEventsHandler._addToFsEvents(path));
   } else {
     if (!this._readyCount) this._readyCount = 0;
