diff --git a/lib/rules/export.js b/lib/rules/export.js
index a155254c8e4c00816ef0cb128ab8c8b6d60be838..8baf0fbd492ad5d34b6e2932afe78bced223cdbc 100644
--- a/lib/rules/export.js
+++ b/lib/rules/export.js
@@ -39,17 +39,21 @@ var tsTypePrefix = 'type:';
 function isTypescriptFunctionOverloads(nodes) {
   var nodesArr = Array.from(nodes);
 
-  var idents = (0, _arrayPrototype2['default'])(
-  nodesArr,
-  function (node) {return node.declaration && (
-    node.declaration.type === 'TSDeclareFunction' // eslint 6+
-    || node.declaration.type === 'TSEmptyBodyFunctionDeclaration' // eslint 4-5
-    ) ?
-    node.declaration.id.name :
-    [];});
-
-  if (new Set(idents).size !== idents.length) {
-    return true;
+  if (nodesArr[0].type === "ExportDefaultDeclaration") {
+    let num = 0;
+    for (const node of nodesArr) {
+      if (
+        // eslint 6+
+        node.declaration.type === 'TSDeclareFunction' ||
+        // eslint 4-5
+        node.declaration.type === 'TSEmptyBodyFunctionDeclaration'
+      ) {
+        num++;
+      }
+    }
+    if (num === nodesArr.length - 1) {
+      return true;
+    }
   }
 
   var types = new Set(nodesArr.map(function (node) {return node.parent.type;}));
