diff --git a/dist/rollup-plugin-dts.mjs b/dist/rollup-plugin-dts.mjs
index f185303864c0df93ec85031163d216d94f27a53b..8bd955f54415a8d69c56bcb5eb01db74ecb79142 100644
--- a/dist/rollup-plugin-dts.mjs
+++ b/dist/rollup-plugin-dts.mjs
@@ -417,9 +417,10 @@ function createReference(id) {
     };
 }
 function createIdentifier(node) {
+    const text = node.getText();
     return withStartEnd({
         type: "Identifier",
-        name: node.getText(),
+        name: text.startsWith(`"`) ? JSON.parse(text) : text,
     }, node);
 }
 /**
