// es6-transpiler/build/es5/lib/scope.js
"use strict";

var assert = require("assert");
var stringmap = require("stringmap");
var stringset = require("stringset");
var is = require("simple-is");
var fmt = require("simple-fmt");
var error = require("./error");

var UUID = 1;

function Scope(args) {
    assert(is.someof(args.kind, ["hoist", "block", "catch-block"]));
    assert(is.object(args.node));
    assert(args.parent === null || is.object(args.parent));

	this.uuid = UUID++;

    // kind === "hoist": function scopes, program scope, injected globals
    // kind === "block": ES6 block scopes
    // kind === "catch-block": catch block scopes
    this.kind = args.kind;

    // the AST node the block corresponds to
    this.node = args.node;

    // parent scope
    this.parent = args.parent;

    // this Scope is a wrapper. wrapper scopes will be skipped from hoist scope searching
    this.wrapper = !!args.wrapper;

    // children scopes for easier traversal (populated internally)
    this.children = [];

    // scope declarations. decls[variable_name] = {
    //     kind: "fun" for functions,
    //           "param" for function parameters,
    //           "caught" for catch parameter
    //           "var",
    //           "const",
    //           "let"
    //     node: the AST node the declaration corresponds to
    //     from: source code index from which it is visible at earliest
    //           (only stored for "const", "let" [and "var"] nodes)
    // }
    this.decls = stringmap();

	// variables (probably temporary) that are free-for-change
	this.freeVariables = [];

    // names of all declarations within this scope that was ever written
    // TODO move to decls.w?
    // TODO create corresponding read?
    this.written = stringset();

    // names of all variables declared outside this hoist scope but
    // referenced in this scope (immediately or in child).
    // only stored on hoist scopes for efficiency
    // (because we currently generate lots of empty block scopes)
    this.propagates = (this.kind === "hoist" ? stringset() : null);

    // scopes register themselves with their parents for easier traversal
    if (this.parent) {
        this.parent.children.push(this);
    }
}

Scope.setOptions = function(options) {
	Scope.options = options;
};

Scope.prototype.print = function(indent) {
    indent = indent || 0;
    var scope = this;
    var names = this.decls.keys().map(function(name) {
        return fmt("{0} [{1}]", name, scope.decls.get(name).kind);
    }).join(", ");
    var propagates = this.propagates ? this.propagates.items().join(", ") : "";
    console.log(fmt("{0}{1}: {2}. propagates: {3}", fmt.repeat(" ", indent), this.node.type, names, propagates));
    this.children.forEach(function(c) {
        c.print(indent + 2);
    });
};

function isObjectPattern(node) {
    return node && node.type === 'ObjectPattern';
}

function isArrayPattern(node) {
    return node && node.type === 'ArrayPattern';
}

function isFunction(node) {
	var type;
	return node && ((type = node.type) === "FunctionDeclaration" || type === "FunctionExpression" || type === "ArrowFunctionExpression");
}

function isLoop(node) {
	var type;
	return node && ((type = node.type) === "ForStatement" || type === "ForInStatement" || type === "ForOfStatement" || type === "WhileStatement" || type === "DoWhileStatement");
}

function isBlockScoped(kind) {
    return is.someof(kind, ["const", "let", "caught"]);
}

Scope.prototype.mutate = function(newKind) {
	if( this.kind !== newKind ) {
		assert(is.someof(newKind, ["hoist", "block", "catch-block"]));

		this.kind = newKind;

		this.propagates = (this.kind === "hoist" ? stringset() : null);
	}

};

Scope.prototype.add = function(name, kind, node, referableFromPos, freeFromPosition, originalDeclarator) {
    assert(is.someof(kind, ["fun", "param", "var", "caught", "const", "let", "get", "set"]), kind + " is a wrong kind");

    var isGlobal = node && (node.range || [])[0] < 0;

    if (!name && isObjectPattern(node)) {
        node.properties.forEach(function(property) {
            property && this.add(property.value.name, kind, property.value, referableFromPos, freeFromPosition, property.key);
        }, this);

        return;
    }
    if (!name && isArrayPattern(node)) {
        node.elements.forEach(function(element) {
            element && this.add(element.name, kind, element, element.range[0])
        }, this);

        return;
    }
	if (!name && node.type === "SpreadElement") {
		name = node.argument.name;
	}

    function isConstLet(kind) {
        return is.someof(kind, ["const", "let"]);
    }

    var scope = this;

    // search nearest hoist-scope for fun, param and var's
    // const, let and caught variables go directly in the scope (which may be hoist, block or catch-block)
    if (is.someof(kind, ["fun", "param", "var"])) {
        while (scope.kind !== "hoist") {
            if (scope.decls.has(name) && isConstLet(scope.decls.get(name).kind)) { // could be caught
                return error(node.loc.start.line, "{0} is already declared", name);
            }
            scope = scope.parent;
        }
    }

    {// name exists in scope and either new or existing kind is const|let|catch => error
        var hasAlready = scope.decls.has(name);
        var existingKind;

        if ( hasAlready ) {
            existingKind = scope.decls.get(name).kind;
        }
        else {// Special test for "catch-block". Maybe it's not so prettily, but it works
            var parentScope = scope.parent;

            if ( parentScope && parentScope.kind === "catch-block" ) {
                hasAlready = parentScope.decls.has(name)
                    && (existingKind = parentScope.decls.get(name).kind) === "caught"
                ;
            }
        }

        if ( hasAlready ) {
            if ( kind == 'get' && existingKind == 'set' || kind == 'set' && existingKind == 'get' ) {
                // all is fine
                //TODO:: do something to allow getter and setter declaration both be in scope.decls
            }
            else if ( Scope.options.disallowDuplicated || isBlockScoped(existingKind) || isBlockScoped(kind) ) {
                return error(node.loc.start.line, "{0} is already declared", name);
            }
        }

    }

    var declaration = {
        kind: kind,
        node: node,
        isGlobal: isGlobal,
        refs: [
            /*reference node's*/
        ]
    };
    if ( referableFromPos !== void 0 ) {
        assert(is.someof(kind, ["var", "const", "let"]), kind + " is not one of [var, const, let]");
        if (originalDeclarator) {
            declaration.from = originalDeclarator.range[0];
        }
        else {
            declaration.from = referableFromPos;
        }
    }
    if ( freeFromPosition !== void 0 ) {
        assert(is.someof(kind, ["var", "const", "let"]), kind + " is not one of [var, const, let]");
        if (originalDeclarator) {
            declaration.to = originalDeclarator.range[1];
        }
        else {
            declaration.to = freeFromPosition;
        }
    }
    scope.decls.set(name, declaration);
};

Scope.prototype.getKind = function(name) {
    assert(is.string(name), "name " + "'" + name + "' must be a string");
    var decl = this.decls.get(name);
    return decl ? decl.kind : null;
};

Scope.prototype.getNode = function(name) {
    assert(is.string(name), "name " + "'" + name + "' must be a string");
    var decl = this.decls.get(name);
    return decl ? decl.node : null;
};

Scope.prototype.getFromPos = function(name) {
    assert(is.string(name), "name " + "'" + name + "' must be a string");
    var decl = this.decls.get(name);
    return decl ? decl.from : null;
};

Scope.prototype.addRef = function(node) {
	assert(node && typeof node === 'object');
	var name = node.name;
	assert(is.string(name), "name " + "'" + name + "' must be a string");
	var decl = this.decls.get(name);
	assert(decl, "could not find declaration for reference " + name);
	decl.refs.push(node);
};

Scope.prototype.getRefs = function(name) {
	assert(is.string(name), "name " + "'" + name + "' must be a string");

	var decl = this.decls.get(name);
	return decl ? decl.refs : null;
};

Scope.prototype.get = function(name) {
    return this.decls.get(name);
};

Scope.prototype.hasOwn = function(name) {
    return this.decls.has(name);
};

Scope.prototype.remove = function(name) {
    return this.decls.delete(name);
};

Scope.prototype.pushFree = function(name, endsFrom) {
	this.freeVariables.push({name: name, endsFrom: endsFrom});
	return name;
};
Scope.prototype.popFree = function(startsFrom) {
	var candidate;

	for( var index = 0, length = this.freeVariables.length ; index < length ; index++ ) {
		candidate = this.freeVariables[index];

		if( candidate.endsFrom && candidate.endsFrom <= startsFrom ) {
			this.freeVariables.splice(index, 1);
			candidate = candidate.name;
			break;
		}

		candidate = null;
	}

	return candidate || null;
};

Scope.prototype.doesPropagate = function(name) {
	assert(this.kind === "hoist");
	return this.kind === "hoist" && this.propagates.has(name);
};

Scope.prototype.markPropagates = function(name) {
	assert(this.kind === "hoist");
	this.propagates.add(name);
};

Scope.prototype.doesThisUsing = function() {
	return this.__thisUsing || false;
};
Scope.prototype.markThisUsing = function() {
	this.__thisUsing = true;
};
Scope.prototype.doesArgumentsUsing = function() {
	return this.__argumentsUsing || false;
};
Scope.prototype.markArgumentsUsing = function() {
	this.__argumentsUsing = true;
};

Scope.prototype.closestHoistScope = function() {
    var scope = this;
    while ( scope.kind !== "hoist" || scope.wrapper ) {
        scope = scope.parent;
    }
    return scope;
};

Scope.prototype.lookup = function(name) {
    for (var scope = this; scope; scope = scope.parent) {
        if (scope.decls.has(name)
            || isFunction(scope.node)
                && scope.node.rest
                && scope.node.rest.name == name
        ) {
            return scope;
        } else if (scope.kind === "hoist") {
            scope.propagates.add(name);
        }
    }
    return null;
};

Scope.prototype.markWrite = function(name) {
    assert(is.string(name), "name " + "'" + name + "' must be a string");
    this.written.add(name);
};

// detects let variables that are never modified (ignores top-level)
Scope.prototype.detectUnmodifiedLets = function() {
    var outmost = this;

    function detect(scope) {
        if (scope !== outmost) {
            scope.decls.keys().forEach(function(name) {
                if (scope.getKind(name) === "let" && !scope.written.has(name)) {
                    return error(scope.getNode(name).loc.start.line, "{0} is declared as let but never modified so could be const", name);
                }
            });
        }

        scope.children.forEach(function(childScope) {
            detect(childScope);;
        });
    }
    detect(this);
};

Scope.prototype.traverse = function(options) {
    options = options || {};
    var pre = options.pre;
    var post = options.post;

    function visit(scope) {
        if (pre) {
            pre(scope);
        }
        scope.children.forEach(function(childScope) {
            visit(childScope);
        });
        if (post) {
            post(scope);
        }
    }

    visit(this);
};

//function __show(__a, scope) {
//	__a += " " ;
//
//	if( scope ) {
//		console.log(__a + " | " + scope.uuid);
//
//		if( (scope.children[0] || {}).uuid == scope.uuid ) {
//			throw new Error("123")
//		}
//		scope.children.forEach(__show.bind(null, __a))
//	}
//}
//
//Object.keys(Scope.prototype).forEach(function(name) {//wrap all
//	var fun = Scope.prototype[name];
//	Scope.prototype[name] = function() {
//		try {
//			return fun.apply(this, arguments);
//		}
//		catch(e) {
//			console.error("ERROR:", name)
//			__show(" ", this);
//		}
//	}
//})

module.exports = Scope;
