// es6-transpiler/build/es5/transpiler/spread.js
"use strict";

var assert = require("assert");
var error = require("./../lib/error");
var core = require("./core");

function findSpreadArgument(node) {
	var spreadIndex = -1;
	var elements;
	if( node && Array.isArray(elements = (node.elements || node["arguments"])) ) {
		elements.forEach(function(arg, i) {
			if( spreadIndex === -1 && core.is.isSpreadElement(arg) ) {
				spreadIndex = i;
			}
		});
	}
	return spreadIndex;
}

function getRange(node) {
	return node && (node.bracketsRange || node.groupRange || node.range);
}

function needFirstSquareBracket(args) {
	var argsLength = args.length;
	return argsLength > 1
		|| (
			core.is.isArrayExpression(args[0].argument)
			&& !core.is.isSpreadElement(args[0].argument.elements[0])
		)
	;
}

function getLastNotNullElementIndex(elements, index) {
	if ( index === void 0 )index = elements.length;

	var element = null;

	while( element == null && index > 0 ) {
		element = elements[--index];
	}

	return element === null ? -1 : index;
}
function createArrayHoles(holesCount) {
	return holesCount > 0
		? "(new Array(" + holesCount + "))"
		: ""
}

var plugin = module.exports = {
	reset: function() {

	}

	, setup: function(alter, ast, options) {
		if( !this.__isInit ) {
			this.reset();
			this.__isInit = true;
		}

		this.alter = alter;
		this.options = options;
	}

	, ':: CallExpression,NewExpression,ArrayExpression': function(node) {
		var spreadIndex = findSpreadArgument(node);
		if( spreadIndex !== -1 ) {
			var type = node.type;

			if ( type === "CallExpression" ) {
				this.replaceCallExpression(node);
			}
			else if ( type === "NewExpression" ) {
				this.replaceNewExpression(node);
			}
			else if ( type === "ArrayExpression" && node.$despread !== true ) {
				this.replaceArrayExpression(node);
			}
		}
	}

	, replaceCallExpression: function(node) {
		var isMemberExpression = node.callee.type === "MemberExpression";
		var calleeType = isMemberExpression && node.callee.object.type;
		var isSimpleMemberExpression = isMemberExpression && (calleeType === "Identifier" || calleeType === "ThisExpression");
		var args = node["arguments"];
		var argsLength = args.length;
		var isSuper = isMemberExpression
			? node.callee.object.$originalName === 'super' || node.callee.object.name === 'super'
			: node.callee.$originalName === 'super' || node.callee.name === 'super'
		;

		assert(argsLength);

		var expressionBefore = ""
			, expressionInside = ""
		;

		if( isMemberExpression ) {
			if( isSimpleMemberExpression || isSuper ) {
				if ( isSuper ) {
					expressionInside =
						".apply(this, "
					;
				}
				else {
					expressionInside =
						".apply("
							+ this.alter.get(node.callee.object.range[0], node.callee.object.range[1])
							+ ", "
					;
				}
			}
			else {
				var tempVar = core.getScopeTempVar(node.callee.object, node.$scope);

				this.alter.wrap(
					node.callee.object.range[0]
					, node.callee.object.range[1]
					, (core.detectSemicolonNecessity(node) ? ";" : "") + "(" + tempVar + " = "
					, ")"
//					, {extend: true}
				);

				expressionInside = ".apply(" + tempVar + ", ";

				core.setScopeTempVar(tempVar, node, node.$scope, true);
			}
		}
		else {
			if ( isSuper ) {
				expressionInside =
					".apply(this, "
				;
			}
			else {
				expressionInside =
					".apply(null, "
				;
			}

			if( node.callee.type === "FunctionExpression" ) {
				expressionBefore = "(";
				expressionInside = ")" + expressionInside;
			}
		}

		if( expressionBefore ) {
			this.alter.insertBefore(
				node.callee.range[0]
				, expressionBefore
			);
		}

		this.alter.replace(
			node.parenthesesRange[0]
			, getRange(args[0])[0]
			, expressionInside + (needFirstSquareBracket(args) ? "[" : "")
		);

		this.replaceSpreads(node, args, true);
	}

	, replaceNewExpression: function(node) {
		// http://jsperf.com/object-create-bind-aply/2

		var isMemberExpression = node.callee.type === "MemberExpression";
		var isFunctionExpression = node.callee.type === "FunctionExpression";
		var args = node["arguments"];
		var argsLength = args.length;
		var isSuper = isMemberExpression
			? node.callee.object.$originalName === 'super' || node.callee.object.name === 'super'
			: node.callee.$originalName === 'super' || node.callee.name === 'super'
		;

		assert(argsLength);

		this.alter.remove(node.range[0], node.range[0] + 3);// remove 'new'

		var tempVarInstance = core.getScopeTempVar(node, node.$scope);
		var tempVar = core.getScopeTempVar(node, node.$scope);

		this.alter.insert(
			node.range[0]
			, (core.detectSemicolonNecessity(node) ? ";" : "")
				+ '((' + tempVar + '='
				+ (isFunctionExpression ? '(' : '')
				+ '((' + tempVarInstance + '=' + core.createVars(node, "create") + '((' + tempVar + '='
		);
		this.alter.insertBefore(
			node.parenthesesRange[0]
			, (isSuper
				? ')===null?null:' + tempVar + '.prototype))'
				: ').prototype))'
				) + ','
				+ (isSuper
					? tempVar + '===null?function(){}:' + tempVar
					: tempVar
					) + ').apply'
		);

		this.alter.replace(
			node.parenthesesRange[0] + 1
			, getRange(args[0])[0]
			, tempVarInstance + ', ' + (needFirstSquareBracket(args) ? '[' : '')
		);

//		this.alter.insertBefore(getRange(args)[1],
		this.alter.insert(node.range[1],
			(isFunctionExpression ? ')' : '')
			+ ')&&typeof ' + tempVar + '===\'object\'?' + tempVar + ':' + tempVarInstance + ')');

		core.setScopeTempVar(tempVarInstance, node, node.$scope, true);
		core.setScopeTempVar(tempVar, node, node.$scope, true);

		this.replaceSpreads(node, args, true);
	}

	, replaceArrayExpression: function(node) {
		// found new line symbols
		this.replaceSpreads(node);
	}

	, replaceSpreads: function(node, elements, innerSpread) {;var $that$0=this;
		var self = this;

		var nonSpreadElementStart = -1;
		function startNotSpreadElementsGroup(start) {
			if( nonSpreadElementStart === -1 ) {
				nonSpreadElementStart = start;
			}
		}
		function endNotSpreadElementsGroup(end) {
			if( nonSpreadElementStart < 0 )return;

			var startElement = elements[nonSpreadElementStart]
				, endElement = elements[end || elements.length - 1]
			;

			assert(!!startElement);

			if( !endElement ) {
				endElement = elements[getLastNotNullElementIndex(elements, end)];
			}

			self.alter
				.insertBefore(getRange(startElement)[0], "[")
				.insert(getRange(endElement)[1], "]")
			;

			nonSpreadElementStart = -1;
		}

		var firstElemIsSpreadFlag = -1, that = this;

		function removeFirstSquareBracket() {
			if ( firstElemIsSpreadFlag >= 0 ) {
				if( !concatOpen ) {
					var from = firstElemIsSpreadFlag;

					that.alter.remove(from, from + 1);
				}
				firstElemIsSpreadFlag = -1;
			}
		}

		elements = elements || node.elements;

		var argsLength = elements.length;
		var callIteratorOpen = false, concatOpen = false, prevConcatOpen = false;
		var lastNullElementCount = 0, ArrayExpressionElementInInnerSpread = null;
		var callIteratorFunctionName, endString = "";
		var i = 0;

		for( ; i < argsLength ; i++ ) {(function(){
			var element = elements[i]
				, elementRange = getRange(element)
			;
			var isFirst = i === 0
				, isLast = i === argsLength - 1
				, isLastAndNull = isLast && element == null
				, theOnlyOne = isLast && isFirst
				, isSpreadEl = core.is.isSpreadElement(element)
			;

			prevConcatOpen = concatOpen;

			if( concatOpen ) {
				if( element || isLastAndNull ) {
					if( lastNullElementCount ) {
						var fistNullIndex = i - lastNullElementCount;

						if( isLastAndNull )lastNullElementCount++;// [1, ...a, , , , ] -> [1].concat(ITER(a), [ , , , ]) case

						if( !isSpreadEl ) {/* see: LINK createArrayHoles#1 */
							$that$0.alter.replace(
								getRange(elements[getLastNotNullElementIndex(elements, fistNullIndex)])[1]
								, isLastAndNull ? node.range[1] - 1 : elementRange[0]
								, (isFirst ? "" : ", ") + createArrayHoles(lastNullElementCount) + (isLastAndNull ? "" : ", ")
							);
						}

						lastNullElementCount = 0;
					}
				}
				else {
					lastNullElementCount++;
				}
			}

			if( isSpreadEl ) {
				var argument = element.argument
					, argumentRange = getRange(argument)
				;

				if ( core.is.isArrayExpression(argument) ) {
					removeFirstSquareBracket();
					firstElemIsSpreadFlag = theOnlyOne && core.is.isSpreadElement(argument.elements[0]) ? argument.range[0] : -1;
					// link `spread_in_spread` 1
					// cases: [...[...obj]] and [...[N, , ...obj]], [...([...obj])] and [...([N, , ...obj])]
					$that$0.alter.remove(elementRange[0], argument.range[0] + (firstElemIsSpreadFlag === -1 ? 1 : 0), {a: 1});// remove "...[" or "...(["

					if ( argument.groupRange ) {
						// cases: [...([...obj])] and [...([N, , ...obj])]
						$that$0.alter.remove(argument.range[1], argument.groupRange[1], {a: 2});// remove ")" from "])"
					}
					var index;
					if ( !isLast
						&& (index = getLastNotNullElementIndex(argument.elements)) != argument.elements.length - 1
					) {
						// cases: [...[ , ], 1] -> [, 1], [...[2, ], 1] -> [2, 1]
						var start = index == -1 ? argument.range[0] + 1 : argument.elements[index].range[0]
							, end = argument.range[1] - 1
						;
						$that$0.alter.remove(start, end, {a: 3});// remove "," from "[ , ]"
					}

					if( !innerSpread || ArrayExpressionElementInInnerSpread ) {
						// link `spread_in_spread` 2
						// cases: [...[N, , ...obj]]
						$that$0.alter.remove(argument.range[1] - 1, argumentRange[1], {a: 4});// remove "]"
					}
					else {
						// FIXME: do something better
						ArrayExpressionElementInInnerSpread = argument;
					}

					elements.splice.apply(elements, [i, 1].concat(argument.elements));
					i--;
					argsLength = elements.length;

					argument.$despread = true;

					return;
				}
				else {
					endNotSpreadElementsGroup(i - 1);

					if( !callIteratorFunctionName ) callIteratorFunctionName = core.createVars(node, 'callIterator');

					var concatStr = concatOpen ? ", " :
						theOnlyOne
							? ""
							: ( (isFirst ? " " : "") + "].concat(" )
					;
					if( !concatOpen )concatOpen = !!concatStr;

					var callIteratorCloseStr = $that$0.__detectNeedToCloneSpreadVariable(argument, elements, i)
						? ", true)"
						: ")"
					;

					if( isFirst ) {
						// TODO:: one this.alter.replace
						var from = elementRange[0];
						if ( !innerSpread && !concatStr ) {
							do {
								--from;
							}
							while ( $that$0.alter.getRange(from, from + 1) !== "[" );
						}

						$that$0.alter.remove(//remove "[..." or "..."
							from
							, argumentRange[0], {a: 5}
						);
						$that$0.alter.insertBefore(
							argumentRange[0]
							, concatStr + callIteratorFunctionName + "("
						);
						if( !theOnlyOne && (concatOpen || callIteratorCloseStr) ) {
							$that$0.alter.insert(argumentRange[1], callIteratorCloseStr);
						}
					}
					else {
						var lastNotNullElementIndex = getLastNotNullElementIndex(elements, i);

						if( lastNotNullElementIndex === -1 ) {// [ , , ...spread] case
							var arrayHolesString = createArrayHoles(i);
							$that$0.alter.replace(
								node.range[0]
								, argumentRange[0]
								, arrayHolesString + ".concat([" + concatStr + callIteratorFunctionName + "("
							);
							endString = ")";
						}
						else {
							var arrayHolesString$0 = createArrayHoles(i - lastNotNullElementIndex - 1);/* LINK createArrayHoles#1 */
							if( arrayHolesString$0 ) {
								arrayHolesString$0 += ", ";
							}

							var lastElement = elements[lastNotNullElementIndex];
							var from$0 = getRange(lastElement)[1]
								, to = argumentRange[0]
							;

							var lineBreaks = $that$0.alter.getRange(from$0, to).match(/[\r\n]/g) || [];
							var lineBreaksCount = lineBreaks.length;

							$that$0.alter.replace(
								from$0
								, to
								, $that$0.alter.get(from$0, to) + '|' + concatStr + arrayHolesString$0 + callIteratorFunctionName + "("
								, {transform: function(str) {
									// HACK START: fix strange bug: removing last ')' if we come from replaceNewExpression
									var char0, add = '', index = 0;
									do {
										char0 = str.charAt(index);
										index++;
										if ( char0 === ')' ) {
											add += ')';
										}
									}
									while ( char0 != '|' );
									str = add + str.substr(index);
									// HACK END

									var newLineBreaks = str.match(/[\r\n]/g) || [];
									var newLineBreaksCount = newLineBreaks.length;

									if ( newLineBreaksCount < lineBreaksCount ) {
										str = lineBreaks.slice(newLineBreaksCount).join("") + str;
									}
									return str;
								}}
							);
						}
						$that$0.alter.insert(argumentRange[1], callIteratorCloseStr);
					}

					callIteratorOpen = true;
				}
			}
			else if( concatOpen ) {
				if( element ) {
					startNotSpreadElementsGroup(i);
				}
				else {
					endNotSpreadElementsGroup(i);
				}
			}

			removeFirstSquareBracket();
		})();}
		endNotSpreadElementsGroup();

		if( endString ) {
			this.alter.insertBefore(node.range[1], endString, {extend: true});
		}

		if ( elements.length == 1 && !concatOpen && callIteratorOpen ) {
			if ( core.is.isArrayExpression(node) ) {// for "NewExpression" and "CallExpression", node is not a SpreadElement )
				this.alter.remove(node.range[0], node.range[0] + 1);
			}
		}

		if( (concatOpen || callIteratorOpen) ) {
			if( innerSpread ) {
				this.alter.insertBefore(node.range[1], ")", {extend: true});
				if( ArrayExpressionElementInInnerSpread ) {
					this.alter.remove(getRange(ArrayExpressionElementInInnerSpread)[1] - 1, getRange(ArrayExpressionElementInInnerSpread)[1], {a: 6});// remove "]"
					ArrayExpressionElementInInnerSpread = null;
				}
			}
			else {
				this.alter.replace(node.range[1] - 1, node.range[1], ")", {extend: true});
			}

		}
	}

	, __detectNeedToCloneSpreadVariable: function(variable, elements, startsFrom) {
		startsFrom = +startsFrom + 1;

		if ( startsFrom >= elements.length )return false;
		if ( core.is.isSequenceExpression(variable) )return true;// TODO: need more work to deep analise SequenceExpression
		if ( variable.type !== "Identifier" )return false;

		elements = elements.slice(startsFrom);

		var variableName = variable.name;

		function checkElement(variableName, element, parentType) {
			var elementType = element.type;

			if( elementType === "Identifier" ) {
				return element.name === variableName
					&& parentType !== "Property"		// [...a, 0, {prop: a}]
					&& parentType !== "ArrayExpression"	// [...a, 0, [a]]
					&& parentType !== "SpreadElement"	// [...a, 0, ...a]
				;
			}
			else if ( core.is.isSpreadElement(element) || core.is.isArgumentExpression(element) ) {
				return checkElement.call(this, variableName, element.argument, element.type);
			}
			else if( elementType === "BinaryExpression" ) {
				return checkElement.call(this, variableName, element.left)
					|| checkElement.call(this, variableName, element.right)
				;
			}
			else if( elementType === "MemberExpression" ) {
				return checkElement.call(this, variableName, element.object);
			}
			else if( elementType === "Property" ) {
				return checkElement.call(this, variableName, element.value, elementType);
			}
			else if( elementType === "ArrayExpression" ) {
				return checkManyElements.call(this, variableName, element.elements, elementType);
			}
			else if( elementType === "SequenceExpression" ) {
				return checkManyElements.call(this, variableName, element.expressions, elementType);
			}
			else if( elementType === "ObjectExpression" ) {
				return /*parentType !== "ArrayExpression" || */checkManyElements.call(this, variableName, element.properties, elementType);
			}
			else if( elementType === "CallExpression" || elementType === "YieldExpression" ) {
				return true;
			}
			else if( elementType === "Literal" ) {
				return false;
			}
			else {
				// all other nodeS can affect to the original variable. Example: var arr = [1], a = [...arr, (()=>(arr.push(2),arr))()]
				return parentType !== "Property" && parentType !== "ArrayExpression";
			}
		}

		function checkManyElements(variableName, elements, parentType) {
			return elements.some(function(element) {
				return element && checkElement.call(this, variableName, element, parentType);
			}, this);
		}

		return checkManyElements(variableName, elements);
	}
};

for(var i in plugin) if( plugin.hasOwnProperty(i) && typeof plugin[i] === "function" ) {
	plugin[i] = plugin[i].bind(plugin);
}
