// astquery/build/es5/keys/es5.js
module.exports = {
	ArrayExpression: ['elements'],
	AssignmentExpression: ['left', 'right'],
	BlockStatement: ['body'],
	BinaryExpression: ['left', 'right'],
	BreakStatement: ['label'],
	CallExpression: ['callee', 'arguments'],
	CatchClause: ['param', 'body'],
	ConditionalExpression: ['test', 'consequent', 'alternate'],
	ContinueStatement: ['label'],
	DebuggerStatement: [],
	DirectiveStatement: [],
	DoWhileStatement: ['body', 'test'],
	EmptyStatement: [],
	ExpressionStatement: ['expression'],
	ForStatement: ['init', 'test', 'update', 'body'],
	ForInStatement: ['left', 'right', 'body'],
	FunctionDeclaration: ['id', 'params', 'body'],
	FunctionExpression: ['id', 'params', 'body'],
	Identifier: [],
	IfStatement: ['test', 'consequent', 'alternate'],
	Literal: [],
	LabeledStatement: ['label', 'body'],
	LogicalExpression: ['left', 'right'],
	MemberExpression: ['object', 'property'],
	NewExpression: ['callee', 'arguments'],
	ObjectExpression: ['properties'],
	Program: ['body'],
	Property: ['key', 'value'],
	ReturnStatement: ['argument'],
	SequenceExpression: ['expressions'],
	SwitchStatement: ['discriminant', 'cases'],
	SwitchCase: ['test', 'consequent'],
	ThisExpression: [],
	ThrowStatement: ['argument'],
	TryStatement: ['block', 'handlers', 'handler', 'guardedHandlers', 'finalizer'],
	UnaryExpression: ['argument'],
	UpdateExpression: ['argument'],
	VariableDeclaration: ['declarations'],
	VariableDeclarator: ['id', 'init'],
	WhileStatement: ['test', 'body'],
	WithStatement: ['object', 'body']
};
