// astquery/build/es5/keys/es6.js
module.exports = {
	ArrayExpression: ['elements'],
	ArrayPattern: ['elements'],
	ArrowFunctionExpression: ['params', 'defaults', 'body', 'rest'],
	AssignmentExpression: ['left', 'right'],
	BinaryExpression: ['left', 'right'],
	BlockStatement: ['body'],
	BreakStatement: ['label'],
	CallExpression: ['callee', 'arguments'],
	CatchClause: ['param', 'body'],
	ClassBody: ['body'],
	ClassDeclaration: ['id', 'body', 'superClass'],
	ClassExpression: ['id', 'body', 'superClass'],
	ComprehensionBlock: ['left', 'right'],
	ComprehensionExpression: ['filter', 'blocks', 'body'],
	ConditionalExpression: ['test', 'consequent', 'alternate'],
	ContinueStatement: ['label'],
	DebuggerStatement: [],
	DoWhileStatement: ['body', 'test'],
	EmptyStatement: [],
	ExportDeclaration: ['declaration', 'specifiers', 'source'],
	ExportBatchSpecifier: [],
	ExportSpecifier: ['id', 'name'],//TODO:: check 'name' needs
	ExpressionStatement: ['expression'],
	ForInStatement: ['left', 'right', 'body'],
	ForOfStatement: ['left', 'right', 'body'],
	ForStatement: ['init', 'test', 'update', 'body'],
	FunctionDeclaration: ['id', 'params', 'defaults', 'body', 'rest'],
	FunctionExpression: ['id', 'params', 'defaults', 'body', 'rest'],
	Identifier: [],
	IfStatement: ['test', 'consequent', 'alternate'],
	ImportDeclaration: ['specifiers', 'source'],
	ImportSpecifier: ['id', 'name'],//TODO:: check 'name' needs
	LabeledStatement: ['label', 'body'],
	Literal: [],
	LogicalExpression: ['left', 'right'],
	MemberExpression: ['object', 'property'],
	MethodDefinition: ['key', 'value'],
	ModuleDeclaration: ['id', 'source', 'body'],
	NewExpression: ['callee', 'arguments'],
	ObjectExpression: ['properties'],
	ObjectPattern: ['properties'],
	Program: ['body'],
	Property: ['key', 'value'],
	ReturnStatement: ['argument'],
	SequenceExpression: ['expressions'],
	SpreadElement: ['argument'],
	SwitchCase: ['test', 'consequent'],
	SwitchStatement: ['discriminant', 'cases'],
	TaggedTemplateExpression: ['tag', 'quasi'],
	TemplateElement: [],
	TemplateLiteral: ['expressions', 'quasis'],
	ThisExpression: [],
	ThrowStatement: ['argument'],
	TryStatement: ['block', 'handlers', 'handler', 'guardedHandlers', 'finalizer'],
	UnaryExpression: ['argument'],
	UpdateExpression: ['argument'],
	VariableDeclaration: ['declarations'],
	VariableDeclarator: ['id', 'init'],
	WhileStatement: ['test', 'body'],
	WithStatement: ['object', 'body'],
	YieldExpression: ['argument']
};
